src/utilities/filetypes/AlfalfaJSON.cpp

```diff
@@ -0,0 +1,223 @@
+#include <utilities/idd/OS_Output_Meter_FieldEnums.hxx>
+#include <utilities/idd/EnergyManagementSystem_OutputVariable_FieldEnums.hxx>
+#include <utilities/idd/EnergyManagementSystem_GlobalVariable_FieldEnums.hxx>
+#include <utilities/idd/EnergyManagementSystem_Actuator_FieldEnums.hxx>
+#include <utilities/idd/Output_Variable_FieldEnums.hxx>
+#include <utilities/idd/Output_Meter_FieldEnums.hxx>
+#include <utilities/idd/IddEnums.hxx>
+
+#include <iostream>
+namespace openstudio {
+namespace alfalfa {
+  namespace detail {
+    AlfalfaJSON_Impl::AlfalfaJSON_Impl() {}
+
+    AlfalfaJSON_Impl::AlfalfaJSON_Impl(const std::string& s) {}
+
+    AlfalfaJSON_Impl::AlfalfaJSON_Impl(const openstudio::path& p) {}
+
+    void AlfalfaJSON_Impl::exposePoint(const AlfalfaPoint& point) {
+      m_points.push_back(point);
+    }
+
+    std::vector<AlfalfaPoint> AlfalfaJSON_Impl::getPoints() {
+      return m_points;
+    }
+
+    bool AlfalfaJSON_Impl::saveAs(const openstudio::path& path) {
+      m_JSONPath = path;
+      return save();
+    }
+
+    bool AlfalfaJSON_Impl::save() const {
+      if (makeParentFolder(m_JSONPath)) {
+        Json::Value root = toJSON();
+        std::ofstream outFile(openstudio::toSystemFilename(m_JSONPath));
+
+        if (outFile) {
+          try {
+            outFile << root;
+            outFile.close();
+            return true;
+          } catch (...) {
+            LOG(Error, "Unable to write file to path '" << toString(m_JSONPath) << "'.");
+            return false;
+          }
+        }
+      }
+      LOG(Error, "Unable to write file to path '" << toString(m_JSONPath) << "', because parent directory "
+                                                  << "could not be created.");
+
+      return false;
+    }
+
+    Json::Value AlfalfaJSON_Impl::toJSON() const {
+      Json::Value root;
+      for (AlfalfaPoint point : m_points) {
+        // No guard here as the toJSON call will throw an exception if the id does not exist.
+        root[point.id().get()] = point.toJSON();
+      }
+      return root;
+    }
+  }  // namespace detail
+
+  AlfalfaJSON::AlfalfaJSON() : m_impl(std::shared_ptr<detail::AlfalfaJSON_Impl>(new detail::AlfalfaJSON_Impl())) {}
+
+  AlfalfaJSON::AlfalfaJSON(const std::string& s) : m_impl(std::shared_ptr<detail::AlfalfaJSON_Impl>(new detail::AlfalfaJSON_Impl(s))) {}
+
+  AlfalfaJSON::AlfalfaJSON(const openstudio::path& p) : m_impl(std::shared_ptr<detail::AlfalfaJSON_Impl>(new detail::AlfalfaJSON_Impl(p))) {}
+
+  boost::optional<AlfalfaPoint> AlfalfaJSON::exposeConstant(float value, const std::string& display_name) {
+    AlfalfaConstant component(value);
+    return exposeFromComponent(component, display_name);
+  }
+
+  boost::optional<AlfalfaPoint> AlfalfaJSON::exposeMeter(const std::string& meter_name, const std::string& display_name) {
+    AlfalfaMeter component(meter_name);
+    return exposeFromComponent(component, display_name);
+  }
```

Comment:
Ok, so if I follow correctly, really the only public class is this AlfalfaJSON class, which exposes a bunch of methods that will:
* Instantiate the derived class of AlfalfaComponent
* Call its own exposeFromComponent with that derived comp..

Reply:
That is the basic pattern, users will mostly be interacting with these helper methods, however, I want to keep the rest of the API available for advanced users to do stuff.